### PR TITLE
Add ability to silence Gem::Package#build

### PR DESCRIPTION
**Pull request by segiddins**
For use in RubyGems.org

## What was the end-user or developer problem that led to this PR?

No way to build a `.gem` without printing to the console or changing global state

## What is your fix for the problem, implemented in this PR?

Add a kwarg to silence the output

## Make sure the following tasks are checked

- [ ] Describe the problem / feature
- [ ] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [ ] Write code to solve the problem
- [ ] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by simi** [comment]
For cli usage there is `--silent` and for "script" usage you should be able to temporary switch ui to silent one. Have you tried something like this?

```ruby
package = Gem::Package.new(...)
package.use_ui(Gem::SilentUI.new) do
  package.build(...)
end
```

**Comment by segiddins** [comment]
That requires changing global state, which isn't ideal for a multiprocess web server, and it also doesn't prevent the string from being constructed

**Comment by simi** [comment]
> That requires changing global state, which isn't ideal for a multiprocess web server, and it also doesn't prevent the string from being constructed

Yes, but that works the same for almost everything inside gem command. Any reason to make this one method special? Any details on the usage of this new argument? Maybe it is worth it to make an exception or explore instance based "ui".

Also if I understand this well, this code change prevents only one specific output of the `build` method, but spec validation can output other (warning, error) messages. :thinking:

**Comment by segiddins** [comment]
Because this is the only one I needed at this timeOn Jun 26, 2023, at 3:23 PM, Josef Šimánek ***@***.***> wrote:﻿

That requires changing global state, which isn't ideal for a multiprocess web server, and it also doesn't prevent the string from being constructed

Yes, but that works the same for almost everything inside gem command. Any reason to make this one method special? Any details on the usage of this new argument? Maybe it is worth it to make an exception or explore instance based "ui".
Also if I understand this well, this code change prevents only one specific output of the build method, but spec validation can output other (warning, error) messages. 🤔

File: lib/rubygems/package.rb [hunk 1]
```diff
@@ -290,7 +290,7 @@ def add_metadata(tar) # :nodoc:
   ##
   # Builds this package based on the specification set by #spec=
 
-  def build(skip_validation = false, strict_validation = false)
+  def build(skip_validation = false, strict_validation = false, silent: false)
```
Comment (deivid-rodriguez) [inline lib/rubygems/package.rb:293]: I also end up avoiding switching from positional to kwargs, just to keep signatures consistent, but in retrospective, we should start to move on at some point 👍.

**Review by deivid-rodriguez — APPROVED** [review]
I'm good with this!

**Comment by simi** [comment]
IMHO this is non-systematic and I still have no clue about your usecase @segiddins to evaluate if that's worth it to merge. :thinking:

**Comment by segiddins** [comment]
https://github.com/rubygems/rubygems.org/pull/3875/files#diff-2252ed8e85781cc8ef901a0b3038fe197bab29ebde207b8b989a53a5af267227R37-R41

**Pull request closed without merging**